odl/deform/optimal_information_transport.py

```diff
@@ -0,0 +1,167 @@
+# Copyright 2014-2016 The ODL development group
+#
+# This file is part of ODL.
+#
+# ODL is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# ODL is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with ODL.  If not, see <http://www.gnu.org/licenses/>.
+
+""":math:`L^p` type discretizations of function spaces."""
+
+# Imports for common Python 2/3 codebase
+from __future__ import print_function, division, absolute_import
+from future import standard_library
+standard_library.install_aliases()
+
+import numpy as np
+
+from odl.discr import Gradient
+from odl.trafos import FourierTransform
+
+__all__ = ('optimal_information_transport_solver',)
+
+
+def optimal_information_transport_solver(op, I, g, niter, eps, sigma, callback=None):
+    DPhiJacobian = op.domain.one()
+
+    grad = Gradient(op.domain, method='central')
+
+    # We solve poisson using the fourier transform
+    ft = FourierTransform(op.domain)
+    k2_values = sum((ft.range.points() ** 2).T)
+    k2 = ft.range.element(np.maximum(np.abs(k2_values), 0.01))
+    poisson_solver = ft.inverse * (1 / k2) * ft
+
+    for _ in range(niter):
+        PhiStarX = DPhiJacobian * I
+
+        tmp = grad(op.adjoint(op(PhiStarX) - g))
+        for i in range(tmp.size):
+            tmp[i] = PhiStarX * tmp[i]
+
+        u = sigma * grad(1 - np.sqrt(DPhiJacobian)) - 2 * tmp
+
+        # solve for v
+        v = grad.range.element()
+        for i in range(u.size):
+            v[i] = poisson_solver(u[i])
+            v[i] -= v[i][0]
+
+        new_points = op.domain.points().T
+        for i in range(tmp.size):
+            new_points[i] -= eps * v[i].ntuple.asarray()
+
```

Comment:
This currently does not provide a way to compute phi inverse, mainly because we have no way of representing it in a simple manner. The way holger discussed should work however, so we could implement that.